lagom/integrations/starlette.py

```diff
@@ -78,13 +78,48 @@ def magic_route(
         :param include_in_schema:
         :return:
         """
-        wrapped_endpoint = self._container.magic_partial(
-            endpoint, shared=self._request_singletons
-        )
+        wrapped = self.wrapped_endpoint_factory(endpoint, self._container.magic_partial)
+
         return Route(
             path,
-            wrapped_endpoint,
+            wrapped,
             methods=methods,
             name=name,
             include_in_schema=include_in_schema,
         )
+
+    def wrapped_endpoint_factory(
+        self, endpoint: Union[Callable, Type[HTTPEndpoint]], partial_provider: Callable
+    ):
+        """Builds an instance of a starlette Route with endpoint callables
+        bound to the container so dependencies can be auto injected.
+
+        :param endpoint:
+        :param partial_provider:
+        """
+        si = self
+
+        if not (isinstance(endpoint, type) and issubclass(endpoint, HTTPEndpoint)):
+            return partial_provider(endpoint, shared=self._request_singletons)
+
+        class HTTPEndpointProxy(HTTPEndpoint):
+            def __init__(self, scope, receive, send):
+                super().__init__(scope, receive, send)
+                self.endpoint = endpoint(scope, receive, send)
+
+            def __getattribute__(self, name: str):
```

Comment:
I've been thinking about your suggestion ` wrap only attributes named after HTTP methods and otherwise not interfere,`. I think you're right and only wrapping the http methods is better. It means we're interfering less with how the classes are implemented and means there's a single entry point for lagom (which generally fits with how it's used in other places).

I suspect it will also make this code easier to follow as it will be more explicit what is decorated and what isn't.

Reply:
:+1: I've made the change.